File: app/src/main/java/deltazero/amarok/xposed/utils/XPref.java
package deltazero.amarok.xposed.utils;

import com.github.kyuubiran.ezxhelper.Log;

import java.util.Collections;
import java.util.Set;

import de.robv.android.xposed.XSharedPreferences;
import deltazero.amarok.BuildConfig;

public class XPref {
    private static XSharedPreferences xPref;

    public static final String XPREF_PATH = "deltazero.amarok.xposed.prefs";
    public static final String HIDE_PKG_NAMES = "hidePkgNames";
    public static final String IS_ACTIVE = "isActive";

    private static boolean isActiveCache = false;
    private static Set<String> hidePkgNamesCache = Collections.emptySet();

    public static void init() {
        Log.d("Initializing XPref...", null);

        xPref = new XSharedPreferences(BuildConfig.APPLICATION_ID, XPREF_PATH);
        Log.d("xPref path: " + xPref.getFile().getAbsolutePath(), null);

        if (xPref.getFile().canRead()) {
            xPref.reload();
            Log.d("xPref content: " + xPref.getAll(), null);
        } else {
            Log.wx("No XPref found. Launch Amarok once to activate XHide.", null);
        }

        Log.ix("XPref initialized.", null);
    }

    public static void refreshCache() {
        xPref.reload();
        if (!xPref.getFile().canRead()) return;
        isActiveCache = xPref.getBoolean(IS_ACTIVE, false);
        hidePkgNamesCache = xPref.getStringSet(HIDE_PKG_NAMES, Collections.emptySet());
    }

    public static boolean isXHideActive() {
        return isActiveCache;
    }

    public static boolean shouldHide(String pkgName) {
        return hidePkgNamesCache.contains(pkgName);
    }

    public static String getXPrefDir() {
        // This function can be called without initializing XPref
        return new XSharedPreferences(BuildConfig.APPLICATION_ID, XPREF_PATH)
                .getFile().getParentFile().getAbsolutePath();
    }
}

File: app/src/main/java/deltazero/amarok/xposed/utils/XPref.kt
package deltazero.amarok.xposed.utils

import com.github.kyuubiran.ezxhelper.Log
import de.robv.android.xposed.XSharedPreferences
import deltazero.amarok.BuildConfig
import java.util.Collections

object XPref {
  private var xPref: XSharedPreferences? = null

  const val XPREF_PATH: String = "deltazero.amarok.xposed.prefs"
  const val HIDE_PKG_NAMES: String = "hidePkgNames"
  const val IS_ACTIVE: String = "isActive"

  private var isActiveCache = false
  private var hidePkgNamesCache: Set<String> = Collections.emptySet()

  @JvmStatic
  fun init() {
    Log.d("Initializing XPref...", null)

    xPref = XSharedPreferences(BuildConfig.APPLICATION_ID, XPREF_PATH)
    Log.d("xPref path: " + xPref!!.file.absolutePath, null)

    if (xPref!!.file.canRead()) {
      xPref!!.reload()
      Log.d("xPref content: " + xPref!!.all, null)
    } else {
      Log.wx("No XPref found. Launch Amarok once to activate XHide.", null)
    }

    Log.ix("XPref initialized.", null)
  }

  @JvmStatic
  fun refreshCache() {
    xPref!!.reload()
    if (!xPref!!.file.canRead()) return
    isActiveCache = xPref!!.getBoolean(IS_ACTIVE, false)
    hidePkgNamesCache =
      xPref!!.getStringSet(HIDE_PKG_NAMES, Collections.emptySet()) ?: Collections.emptySet()
  }

  @JvmStatic fun isXHideActive(): Boolean = isActiveCache

  @JvmStatic fun shouldHide(pkgName: String?): Boolean = hidePkgNamesCache.contains(pkgName)

  @JvmStatic
  fun getXPrefDir(): String =
    // This function can be called without initializing XPref
    XSharedPreferences(BuildConfig.APPLICATION_ID, XPREF_PATH).file.parentFile!!.absolutePath
}
